File: java/src/com/nghiabui/bootstrap/core/Progress.java
package com.nghiabui.bootstrap.core;

import java.util.Collection;
import java.util.List;

public interface Progress {

	void beginLearning(int numWords, int numPatterns);
	void beginIter(int iteration);

	void beginGenNewPatterns(int numWords);
	void endGenNewPatterns(Collection<Pattern> patterns, int numWordsRemain);

	void beginSelectBestPatterns(int numPatterns);
	void endSelectBestPatterns(BestPatterns best);

	void beginExtractNewWords();
	void endExtractNewWords(Collection<Word> words);

	void beginSelectBestWords();
	void endSelectBestWords(List<ScoreRecord<Word>> best);

	void endBootstrap(StopReason reason, Collection<Word> newWords, Collection<Pattern> newPatterns);
	
	void beginPrunning(int numNewPatterns);
	void endPruning(BestPatterns best);
	
}

File: fsharp/Bootstrap.Core/Progress.fs
namespace Bootstrap.Core

open NghiaBui.Common

type FinishReason = EnoughIterations | NoMoreBestPatterns

type Progress =
    abstract BeginLearning              : int -> int -> int -> Param -> unit
    abstract BeginIter                  : int -> unit

    abstract BeginGenNewPatterns        : int -> unit
    abstract EndGenNewPatterns          : IHS<Pattern> -> int -> unit

    abstract BeginSelectBestPatterns    : int -> unit
    abstract EndSelectBestPatterns      : (Pattern * float * bool * bool) [] -> float
                                            -> (Pattern * float * bool * bool) [] -> unit

    abstract BeginExtractNewWords       : unit -> unit
    abstract EndExtractNewWords         : IHS<Word> -> unit

    abstract BeginSelectBestWords       : unit -> unit
    abstract EndSelectBestWords         : (Word * float) [] -> unit

    abstract EndBootstrap               : FinishReason -> IHS<Word> -> IHS<Pattern> -> unit

    abstract BeginPruning               : int -> int -> unit
    abstract EndPruning                 : (Pattern * float) [] -> float
                                            -> (Pattern * float) [] -> (Pattern * float) [] -> unit

    abstract BeginFinalExtract          : unit -> unit
    abstract EndLearning                : IHS<Word> -> IHS<Word> -> IHS<Word> -> unit
